icefoganalytics/elcc-data-management
# ELCC-20: Simplify CentreDashboardWorksheet Loading

Fixes https://yg-hpw.atlassian.net/browse/ELCC-20

# Context

Currently this page is loading all worksheets for all years. It would make the code much easier to read, use, and also speed it up greatly, if this only loaded the worksheet that you were looking at.

e.g. 
the front-end route would become

- `/child-care-centres/:id/fiscal-year/:fiscalYear/:tab/:subTab` 
On the "worksheets" -> Apri tab this would send a query to the back end of the form `/api/funding-submission-line-jsons?where[centreId]=1&where[fiscalYear]=2023/24&where[dateName]=April`

Each component on the page should only load the objects it needs visible, and deep for loops should be avoided for readability.

# Implementation

1. Simplify code in CenterDashboardWorksheet. Remove a lot of it, and push some code down a layer.
2. Vastly speed up worksheet loading by only using v-text-inputs for editable fields in SectionTables. Speed went from ~950ms to around ~350ms. More importantly the code is 10x more readable, and predictable.
3. Standardize fiscal period wiring and patterns between front- and back-ends. i.e. Switch to using latest Controller pattern, and switch from store to use composable for better loading encapsulation.
4. Simplify fiscal period api, and do date parsing at the last mile, rather than in the API.
5. Standardize funding submission line json wiring and patterns between front- and back-ends.
6. Upgrade front-end typescript and vue-tsc packages to fix some type check errors.
7. Fix front-end linting and upgrade linting packages.
8. Add `dev` helper to get Jira Issue description as Markdown.

# Screenshots

New Render Speed
![image](https://github.com/user-attachments/assets/14b9a633-2b4e-4208-be64-9f8b785a3b25)

Old Render Speed
![image](https://github.com/user-attachments/assets/7fe598ff-5b01-4760-9c15-be04612df833)

# Testing Instructions

1. Boot the app via `dev up`
2. Check that the app complies, and that you can log in at http://localhost:8080.
3. Click on the Education is Currently Funding card.
4. Double click on a child care center.
5. Click on the Worksheets tab.
6. If there are no worksheets click the Add Worksheets for xxxx/xx button.
7. Check that the worksheets are added.
8. Enter some data in the first Est column.
9. Tab away, and check that the data propagates down.
10. Check that "save" works, and has less flicker.
11. Check that "replicate" estimates works, and has less flicker.
12. Click on the "May" tab, check that the estimates replicated correctly.
13. Reload the page, check that the data page load a lot faster, not quite real time, but fast enough that the loader barely spins.
